EDGAR/embedder.py

```diff
@@ -0,0 +1,125 @@
+"""
+Embedder class embeds phrases and returns vectors for machine learning processing
+"""
+import pandas as pd
+
+import gensim
+from gensim.models import Word2Vec
+from gensim.test.utils import common_texts
+
+import numpy as np
+
+import re
+import string
+import nltk
+from nltk.stem import PorterStemmer
+import heapq
+
+
+import torch
+from torch.utils.data import TensorDataset, DataLoader, RandomSampler, SequentialSampler
+from pytorch_pretrained_bert import BertTokenizer, BertConfig
+from pytorch_pretrained_bert import BertAdam, BertForSequenceClassification
+
+
+
+"""
+Class embedder takes in three diffeent types: 
+Word2Vec, BagOfWords, BERT 
+""""
+class embedder:
+    def __init__(self, type='BagOfWords', MAX_LEN=100):
+        
+        self.type = type
+        self.MAX_LEN = MAX_LEN
+        
+        if type == 'Word2Vec':
+            pretrained_embeddings_path = 'GoogleNews-vectors-negative300.bin'
+            self.model = gensim.models.KeyedVectors.load_word2vec_format(pretrained_embeddings_path,  binary=True)
+        elif type == 'BagOfWords':
+            self.model = None
+            self.dictionary = {}
+        elif type == 'BERT':
+            self.model = None
+            
+    
+    """ 
+    Cleans phrases by removing non-alpha chracters and stopwords
+    Stemmer is optional, stem='Stem' or None
+    """
+    def clean_phrase(self, text, stem=None):
+        final_string = ""
+        text = text.lower()
+        
+        """ removes non-alpha characters and keeps whitespace """
+        text = re.sub(r'[^A-Za-z0-9 ]', '', text)
+        text = text.split(' ')
+        
+        stopwords = nltk.corpus.stopwords.words("english")
+        # removes stop words and empty strings
+        text_filtered = [word for word in text if not word in stopwords and word != '']
+        if stem == 'Stem':
+            stemmer = PorterStemmer() 
+            text_stemmed = [stemmer.stem(word) for word in text_filtered]
+        else: 
+            text_stemmed = text_filtered
+            
+        final_string = ' '.join(text_stemmed)
+        return final_string
+
+    
+    """
+    Only for Bag Of Words 
+    Adds words to dictionary
+    """
+    def add_dictionary(self, phrase):
+        
+        for word in phrase.split(' '):
+            if word not in self.dictionary.keys():
+                self.dictionary[word] = 1
+            else:
+                self.dictionary[word] += 1
+                
+     
```

Comment:
We need another function here.

```add_dictionary``` isn't really enough by itself to actually use the bag of words model.
We need something like 
```
def generate_vocab_for_corpus(self, texts: List[str], m):
    For text in texts:
        self.add_dictionary(self, text)
    For word in self.vocab:
        if self.vocab[word] (total observed counts) < min_threshold:
            del self.vocab[word]
    # TODO add logic to CAP maximum number of words if user provides maximum_number_words
```